File: math/fibonacci/number.js


// dymaic programming array
var fib = function(n) {

	var dp = [ 0,1 ];
	while(dp[n] == null) dp[dp.length] = dp[dp.length-1] + dp[dp.length-2];
	return dp[n]
}

// recursion
var fib = function(n) {
	return n === 0 ? 0 : n === 1 ? 1 : fib(n-1) + fib(n-2)
}


// iterative
var fib = function(n) {

	if(n === 0) return 0;

	var a = 0;
	var b = 1;
	var m = 2;

	while(m++ <= n) [ a,b ] = [ b,b +a ];
	return b
}

File: math/fibonacci/number.py


# dymaic programming array
def fib(n: int) -> int:

	dp = [ 0,1 ]
	for _ in range(n -1): dp.append(dp[-1] + dp[-2])
	return dp[n]


# recursion
def fib(n: int) -> int:
	return 0 if n == 0 else 1 if n == 1 else fib(n -1) + fib(n -2)


# iterative
def fib(n: int) -> int:

	a,b = 0,1
	for _ in range(n -1): a,b = b,b+a
	return b if n else a

